### PR TITLE
[SPARK-50243][SQL][Connect] Cached classloader for ArtifactManager

### PR DESCRIPTION
**Pull request by xupefei**
### What changes were proposed in this pull request?

This PR implements a caching mechanism for the classloader of `ArtifactManager`, to avoid re-generating a new one every time when a SQL query runs. This change also fixes a longstanding bug where the codegen cache was broken for Spark Connect, due to new classloaders causing cache misses: https://github.com/apache/spark/blob/05728e4ff64e6684d7c6501f8a079e3b9aded9ed/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala#L1487-L1490.

The approach we use is to cache the generated classloader until a new artifact is added.

### Why are the changes needed?

To improve performance and fix codegen caching.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added two new test cases.
Also, some existing tests will fail when the config `ARTIFACTS_SESSION_ISOLATION_ALWAYS_APPLY_CLASSLOADER` is enabled:

- AdaptiveQueryExecSuite: `SPARK-37753: Inhibit broadcast in left outer join when there are many empty partitions on outer/left side`
- DatasetOptimizationSuite: `SPARK-27871: Dataset encoder should benefit from codegen cache`

These tests will not fail after this PR.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala [hunk 1]
```diff
@@ -88,6 +88,10 @@ class ArtifactManager(session: SparkSession) extends Logging {
    */
   protected val sessionArtifactAdded = new AtomicBoolean(false)
 
+  protected val cachedClassLoader = new ThreadLocal[ClassLoader] {
+    override def initialValue(): ClassLoader = null
+  }
+
```
Comment (vicennial) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:94]: I'm not sure I understand why this is a ThreadLocal? Artifacts can be added from multiple threads so wouldn't the addition on one thread cause all the others cached classloaders to be invalidated?

Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala]: > the addition on one thread cause all the others cached classloaders to be invalidated

No, in this case, each thread will have its own classloader, and CodeGen will happen once for each thread (after which will be cached)

Good point. After checking codes about the CodeGen cache I believe it's safe to share classloaders across threads.

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala [hunk 2]
```diff
@@ -282,10 +286,16 @@ class ArtifactManager(session: SparkSession) extends Logging {
     }
   }
 
+  def classloader: ClassLoader = cachedClassLoader.getOrElse {
+    val loader = buildClassLoader
+    cachedClassLoader = Some(loader)
+    loader
+  }
```
Comment (vicennial) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:293]: Synchronized block here to protect against race conditions

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala [hunk 3]
```diff
@@ -88,6 +88,8 @@ class ArtifactManager(session: SparkSession) extends Logging {
    */
   protected val sessionArtifactAdded = new AtomicBoolean(false)
 
+  protected var cachedClassLoader: Option[ClassLoader] = None
```
Comment (vicennial) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:91]: `volatile` to make sure the unset is visible across threads

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala (hunk 2 above)
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala]: Done!
I am thinking if `AtomicReference` would be a better choice here, but I see it's missing a `compareAndExchange` variant that updates the current value when it is not equal to the witness value.

**Review by vicennial — APPROVED** [review]
LGTM thanks!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 4]
```diff
@@ -749,6 +749,7 @@ class AdaptiveQueryExecSuite
     // so retry several times here to avoid unit test failure.
     eventually(timeout(15.seconds), interval(500.milliseconds)) {
       withSQLConf(
+        SQLConf.ARTIFACTS_SESSION_ISOLATION_ALWAYS_APPLY_CLASSLOADER.key -> "true",
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:752]: Hi, @xupefei and @HyukjinKwon . 

May I ask why this PR touches irrelevant `AdaptiveQueryExecSuite` here?

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:752]: Also, I'm observing this test case failure still in some environments although this PR claims that `These tests will not fail after this PR`.
> Also, some existing tests will fail when the config ARTIFACTS_SESSION_ISOLATION_ALWAYS_APPLY_CLASSLOADER is enabled:
> AdaptiveQueryExecSuite: SPARK-37753: Inhibit broadcast in left outer join when there are many empty partitions on outer/left side
> DatasetOptimizationSuite: SPARK-27871: Dataset encoder should benefit from codegen cache
> These tests will not fail after this PR.

**Comment by dongjoon-hyun** [comment]
When I ran `AdaptiveQueryExecSuite` locally without that patch, it seems okay which means the newly added line could be irrelevant to this PR. I made a follow-up to remove that line. Please let me know if I'm wrong, @xupefei and @HyukjinKwon .

- https://github.com/apache/spark/pull/49374

**Pull request closed without merging**